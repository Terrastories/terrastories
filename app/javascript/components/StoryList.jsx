import React, { Component } from "react";
import PropTypes from "prop-types";
import StoryMedia from "./StoryMedia";
import Filter from "./Filter";
import Sort from "./Sort";
import { List, AutoSizer, CellMeasurer, CellMeasurerCache } from "react-virtualized";

class StoryList extends Component {
  constructor(props) {
    super(props)
    this.cache = new CellMeasurerCache({
      fixedWidth: true,
      defaultHeight: 200
    });
    this._isMounted = false;
  }

  static propTypes = {
    stories: PropTypes.array,
    handleStoriesChanged: PropTypes.func,
    handleFilter: PropTypes.func,
    clearFilteredStories: PropTypes.func,
    onStoryClick: PropTypes.func,
    filterMap: PropTypes.object,
    categories: PropTypes.array,
    activeStory: PropTypes.object,
    filterCategory: PropTypes.string,
    filterItem: PropTypes.string,
    handleFilterCategoryChange: PropTypes.func,
    handleFilterItemChange: PropTypes.func,
    itemOptions: PropTypes.array
  };

  // In React 16.3.0, update method to getSnapshotBeforeUpdate
  componentWillReceiveProps(nextProps) {
    if(this._list){
      this._list.forceUpdateGrid();
    }
    if (nextProps.stories !== this.props.stories) {
      this._list.recomputeRowHeights();
      this._list.recomputeGridSize();
      this._list.forceUpdateGrid();
    }
  }

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  handleClickStory = (story, index) => {
    this.props.onStoryClick(story);
  }

  handleFilter = (category, item) => {
    this.props.handleFilter(category, item);
    this._list.scrollToPosition(0);
  }

  getStorySpeakerNames = story => {
    return story.speakers.map(function(speaker) { return speaker.name }).join(',');
  }

  handleClearFilteredStories = () => {
    this.props.clearFilteredStories();
    this._list.scrollToPosition(0);
  }

  renderStory = ({ key, index, style, parent }) => {
    const story = this.props.stories[index];
    let storyClass = '';
    if (this.props.activeStory && this.props.activeStory.id === story.id) {
      storyClass = `story${index} isActive`
    } else {
      storyClass = `story${index}`;
    }
    const bustCache = () => {
      this.cache.clear(index, 0);
    }
    return (
      <CellMeasurer
        key={key}
        cache={this.cache}
        parent={parent}
        columnIndex={0}
        rowIndex={index}>
        <li
          className={storyClass}
          onClick={_ => this.props.onStoryClick(story)}
          style={style}
        >
          <div className="speakers">
            {(story.speakers.length === 1) &&
                <div>
                  <img src={story.speakers[0].picture_url} alt={story.speakers[0].name} title={story.speakers[0].name} />
                  <p style={{ fontWeight: 'bold' }}>{story.speakers[0].name}</p>
                </div>}
            {(story.speakers.length > 1) &&
              story.speakers.map(speaker => (
                  <img src={speaker.picture_url} alt={speaker.name} title={speaker.name} />
              ))
            }
            {(story.speakers.length > 1) &&
              <p style={{ fontWeight: 'bold' }}> {this.getStorySpeakerNames(story)} </p>
            }
          </div>
          <div className="container">
            <h6 className="title">{ story.permission_level === "restricted" && "ðŸ”’" }{story.title}</h6>
            <p>{story.desc}</p>
            {story.media &&
              story.media.map(file => <StoryMedia file={file} doBustCache={bustCache} key={story.media.id} />)}
          </div>
        </li>
      </CellMeasurer>
    );
  };

  render() {
    return (
      <React.Fragment>
        <div className="card--nav">
          <Filter
            handleFilter={this.handleFilter}
            categories={this.props.categories}
            filterMap={this.props.filterMap}
            clearFilteredStories={this.handleClearFilteredStories}
            filterCategory={this.props.filterCategory}
            filterItem={this.props.filterItem}
            handleFilterCategoryChange={this.props.handleFilterCategoryChange}
            handleFilterItemChange={this.props.handleFilterItemChange}
            itemOptions={this.props.itemOptions}
          />
          <Sort
            stories={this.props.stories}
            handleStoriesChanged={this.props.handleStoriesChanged}
          />
        </div>
        <div className="stories">
          <AutoSizer>
            {({ height, width }) => {
              return (
                <List
                  height={height}
                  width={width}
                  rowCount={this.props.stories.length}
                  deferredMeasurementCache={this.cache}
                  rowHeight={this.cache.rowHeight}
                  rowRenderer={this.renderStory}
                  overscanRowCount={4}
                  ref={list => (this._list = list)}
                />
              );
            }}
          </AutoSizer>
        </div>
      </React.Fragment>
    );
  }
}

export default StoryList;
