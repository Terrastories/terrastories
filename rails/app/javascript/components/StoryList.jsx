import React, { Component } from "react";
import PropTypes from "prop-types";
import Story from "./Story";
import Filter from "./Filter";
import Sort from "./Sort";
import ReactList from 'react-list';

class StoryList extends Component {
  static propTypes = {
    stories: PropTypes.array,
    handleStoriesChanged: PropTypes.func,
    handleFilter: PropTypes.func,
    clearFilteredStories: PropTypes.func,
    onStoryClick: PropTypes.func,
    filterMap: PropTypes.object,
    categories: PropTypes.array,
    activeStory: PropTypes.object,
    filterCategory: PropTypes.string,
    filterItem: PropTypes.string,
    handleFilterCategoryChange: PropTypes.func,
    handleFilterItemChange: PropTypes.func,
    itemOptions: PropTypes.array
  };

  handleClickStory = (story, index) => {
    this.props.onStoryClick(story);
  }

  handleFilterItemChange = (item) => {
    this.props.handleFilterItemChange(item);
    this._list.scrollTo(0);
  }

  handleClearFilteredStories = () => {
    this.props.clearFilteredStories();
    this._list.scrollTo(0);
  }

  handleSort = (option) => {
    this.props.handleStoriesChanged(option);
    this._list.scrollTo(0);
  }

  renderStory = (index, key) => {
    const story = this.props.stories[index];
    let storyClass = '';
    if (this.props.activeStory && this.props.activeStory.id === story.id) {
      storyClass = `story${index} isActive`
    } else {
      storyClass = `story${index}`;
    }
    return (
        <Story
          story={story}
          onStoryClick={this.props.onStoryClick}
          storyClass={storyClass}
          key={key}
        />
    );
  };

  render() {
    return (
      <React.Fragment>
        <div className="card--nav">
          <Filter
            categories={this.props.categories}
            filterMap={this.props.filterMap}
            clearFilteredStories={this.handleClearFilteredStories}
            filterCategory={this.props.filterCategory}
            filterItem={this.props.filterItem}
            handleFilterCategoryChange={this.props.handleFilterCategoryChange}
            handleFilterItemChange={this.handleFilterItemChange}
            itemOptions={this.props.itemOptions}
          />
          <Sort
            stories={this.props.stories}
            handleStoriesChanged={this.props.handleStoriesChanged}
          />
        </div>
        <div className="stories">
          <ReactList
            ref={list => this._list = list }
            itemRenderer={this.renderStory}
            length={this.props.stories.length}
            type='variable'
          />
        </div>
      </React.Fragment>
    );
  }
}

export default StoryList;
