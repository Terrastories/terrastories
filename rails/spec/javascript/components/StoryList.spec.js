import React from 'react';
import { shallow } from 'enzyme';
import StoryList from '../../../app/javascript/components/StoryList';

describe('StoryList Component', () => {
  let wrapper;

  beforeEach(() => {
    global.I18n = {
      t: jest.fn(),
    },
    wrapper = shallow(
      <StoryList />,
    );
  })

  it('Displays correctly', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Filter')).toHaveLength(1);
    expect(wrapper.find('IntroductoryPanel')).toHaveLength(1);
    expect(wrapper.find('Button')).toHaveLength(1);
  })

  it('Clicks the explore button', () => {
    wrapper.find('Button').props.handleClick = global.I18n.t;
    wrapper.find('Button').props.handleClick();

    expect(global.I18n.t).toHaveBeenCalled();
  })
})