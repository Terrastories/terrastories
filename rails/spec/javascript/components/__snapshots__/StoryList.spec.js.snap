// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryList Component Displays correctly 1`] = `
<Fragment>
  <div
    className="card--nav"
  >
    <Filter
      categories={Array []}
      clearFilteredStories={[Function]}
      filterMap={Object {}}
      handleFilterItemChange={[Function]}
    />
  </div>
  <div
    className="card--nav terrastory-info"
  >
    <IntroductoryPanel />
    <Button
      buttonText="EXPLORE"
      buttonType="explore-button"
      handleClick={[Function]}
    />
  </div>
</Fragment>
`;
